### DISCUSSION TITLE
Groonga will do full text search against a TABLE_HASH_KEY _key

### DISCUSSION THREAD

**Discussion by yssrku** [groonga/groonga#1888] [General]
I recently encountered some surprising behaviors of Groonga.

Consider a setup like this
```
table_create Coupons TABLE_HASH_KEY ShortText
column_create Coupons title COLUMN_SCALAR ShortText

load --table Coupons
[
    {"_key": "fr1", "title": "Orange 10% off"},
    {"_key": "fr10", "title": "Apple 20% off"}
]
```

I tried to match `_key` with `match_columns` as was surprised that any _key contains the keyword will match. Also, this is quite slow if the table is big.

This syntax is tempting because there's no need do escape user input, and I thought as there's a dedicated syntax for prefix match, Groonga should do exact match in this case. (plus, TABLE_HASH_KEY only supports exact match)

```
select Coupons --match_columns _key --query fr1 --match_escalation no
[[0,1724147170.556698,0.0001139640808105469],[[[2],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]],[1,"fr1","Orange 10% off"],[2,"fr10","Apple 20% off"]]]]
```

I can get the correct result by using "_key:value" syntax.
```
> select Coupons --query _key:fr1 --match_escalation no
[[0,1724147613.411843,0.0006759166717529297],[[[1],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]],[1,"fr1","Orange 10% off"]]]]
```

Another annoying case is if you want to do a search box which user can either search by title or do exact match by id, which is very common in EC use cases.
```
table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenBigram --normalizer NormalizerAuto
column_create Terms coupon_title_index COLUMN_INDEX|WITH_POSITION Coupons title
```

This doesn't work.
```
select Coupons --match_columns "_key || title" --query fr1 --match_escalation no
[[0,1724206933.75881,0.008759975433349609],[[[0],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]]]]]
```

This does
```
select Coupons --query "_key:fr1 OR title:@fr1" --match_escalation no
```

Yet another weird thing is, if I try to add an index to Coupons._key (because I want to figure out a way to use normal `query`), suddenly the behavior changes.
```
> column_create Terms coupon_key_index COLUMN_INDEX Coupons _key
2024-08-21 11:46:36.289738|n| DDL:260:column_create Terms coupon_key_index
2024-08-21 11:46:36.294249|n| spec:260:update:Terms.coupon_key_index:72(column:index):256(Coupons)
2024-08-21 11:46:36.294300|n| DDL:260:set_source Terms.coupon_key_index Coupons
2024-08-21 11:46:36.294321|n| spec:260:update:Terms.coupon_key_index:72(column:index):256(Coupons)
2024-08-21 11:46:36.294333|n| DDL:256:add_hook:insert Coupons [Terms.coupon_key_index(260)]
2024-08-21 11:46:36.294344|n| spec:256:update:Coupons:48(table:hash_key):0
2024-08-21 11:46:36.294359|n| DDL:256:add_hook:delete Coupons [Terms.coupon_key_index(260)]
2024-08-21 11:46:36.294372|n| spec:256:update:Coupons:48(table:hash_key):0
2024-08-21 11:46:36.300051|i| [ii][builder][fin] removed path: <test.00001046LTqHb>
[[0,1724208396.289354,0.01091384887695312],true]
> select Coupons --match_columns "_key" --query "fr10" --match_escalation no
2024-08-21 11:46:43.488686|i| [object][search][index][key][exact] <Terms.coupon_key_index>
2024-08-21 11:46:43.488711|i| grn_ii_sel > (fr10)
2024-08-21 11:46:43.488789|i| [ii][select] n=2 (fr10)
2024-08-21 11:46:43.488812|i| exact: 0
2024-08-21 11:46:43.488821|i| hits=0
[[0,1724208403.488342,0.0005869865417480469],[[[0],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]]]]]
> select Coupons --match_columns "_key" --query "fr10" --match_escalation no
[[0,1724208419.611918,0.000186920166015625],[[[0],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]]]]]
```

**Comment by komainu8** [comment]
TABLE_HASH_KEY can not only search with exact match but also search with full-text-search.
So, if we use [full-text-search-condition](https://groonga.org/docs/reference/grn_expr/query_syntax.html#full-text-search-condition) as your below case, Groonga return records contains the keyword.

```
table_create Coupons TABLE_HASH_KEY ShortText
column_create Coupons title COLUMN_SCALAR ShortText

load --table Coupons
[
    {"_key": "fr1", "title": "Orange 10% off"},
    {"_key": "fr10", "title": "Apple 20% off"}
]
select Coupons --match_columns _key --query fr1 --match_escalation no
[[0,1724147170.556698,0.0001139640808105469],[[[2],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]],[1,"fr1","Orange 10% off"],[2,"fr10","Apple 20% off"]]]]
```

**Reply by yssrku** [reply]
Thanks for the reply, I think my confusion comes from the docs, which says
```
TABLE_HASH_KEY is fast but it doesn’t support advanced search functions such as common prefix search and predictive search.

TABLE_HASH_KEY is useful for index for exact search such as tag search.
```

So basically, I thought it was possible to make full text search against a TABLE_HASH_KEY acts as exact match, turns out it doesn't work that way.

**Comment by komainu8** [comment]
I creating comment about the floowing case. Would you wait for a moment?


```
table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenBigram --normalizer NormalizerAuto
column_create Terms coupon_title_index COLUMN_INDEX|WITH_POSITION Coupons title
...
```

**Reply by komainu8** [reply]
Probably, The cause of the following query doesn't work is https://github.com/groonga/groonga/issues/1828.

```
select Coupons --match_columns "_key || title" --query fr1 --match_escalation no
[[0,1724206933.75881,0.008759975433349609],[[[0],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]]]]]
```

The occurrence condition of https://github.com/groonga/groonga/issues/1828 as below.

* Search with sequential.
* `--match_columns` has multiple columns.

In this case, you didn't set indext to `_key` column.
Groonga search `_key` column with sequential and `--match_columns` has multiple columns.
Therefore, sequential search against `_key` column doesn't work.

**Comment by komainu8** [comment]
The cause of the following problem is not set `WITH_POSITION` to `Coupons._key`'s index.

```
> column_create Terms coupon_key_index COLUMN_INDEX Coupons _key
2024-08-21 11:46:36.289738|n| DDL:260:column_create Terms coupon_key_index
2024-08-21 11:46:36.294249|n| spec:260:update:Terms.coupon_key_index:72(column:index):256(Coupons)
2024-08-21 11:46:36.294300|n| DDL:260:set_source Terms.coupon_key_index Coupons
2024-08-21 11:46:36.294321|n| spec:260:update:Terms.coupon_key_index:72(column:index):256(Coupons)
2024-08-21 11:46:36.294333|n| DDL:256:add_hook:insert Coupons [Terms.coupon_key_index(260)]
2024-08-21 11:46:36.294344|n| spec:256:update:Coupons:48(table:hash_key):0
2024-08-21 11:46:36.294359|n| DDL:256:add_hook:delete Coupons [Terms.coupon_key_index(260)]
2024-08-21 11:46:36.294372|n| spec:256:update:Coupons:48(table:hash_key):0
2024-08-21 11:46:36.300051|i| [ii][builder][fin] removed path: <test.00001046LTqHb>
[[0,1724208396.289354,0.01091384887695312],true]
> select Coupons --match_columns "_key" --query "fr10" --match_escalation no
2024-08-21 11:46:43.488686|i| [object][search][index][key][exact] <Terms.coupon_key_index>
2024-08-21 11:46:43.488711|i| grn_ii_sel > (fr10)
2024-08-21 11:46:43.488789|i| [ii][select] n=2 (fr10)
2024-08-21 11:46:43.488812|i| exact: 0
2024-08-21 11:46:43.488821|i| hits=0
[[0,1724208403.488342,0.0005869865417480469],[[[0],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]]]]]
> select Coupons --match_columns "_key" --query "fr10" --match_escalation no
[[0,1724208419.611918,0.000186920166015625],[[[0],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]]]]]
```

**Reply by komainu8** [reply]
If I set `WITH_POSITION` to Coupons._key's index, Groonga returns correct result as below.

```
column_create Terms coupon_key_index COLUMN_INDEX|WITH_POSITION Coupons _key

select Coupons --match_columns "_key" --query "fr10" --match_escalation no
[[0,1724234798.774976,0.001313209533691406],[[[1],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]],[2,"fr10","Apple 20% off"]]]]
```

**Reply by komainu8** [reply]
If we execute full-text-search with index, the index must set `WITH_POSITION`.
ref: https://groonga.org/docs/reference/commands/column_create.html#create-an-index-column-for-full-text-search

**Reply by komainu8** [reply]
Also, if we set index against `_key` and `title`, `select Coupons --match_columns "_key || title" --query fr1 --match_escalation no` returns correct result as below.

```
select Coupons --match_columns "_key || title" --query fr1 --match_escalation no
[[0,1724235323.170351,0.001344442367553711],[[[1],[["_id","UInt32"],["_key","ShortText"],["title","ShortText"]],[1,"fr1","Orange 10% off"]]]]
```

**Reply by yssrku** [reply]
Thanks, now I understand.

1. Bare `query` only do full text search
2. full text search only works with `WITH_POSITION` index

So I tried the create another table for exact match on _key, seems working well!
```
table_create TermsExact TABLE_PAT_KEY ShortText --default_tokenizer TokenDelimitNull
column_create TermsExact coupon_key_index COLUMN_INDEX|WITH_POSITION Coupons _key
```

**Comment by otegami** [comment]
@yssrku 

Thank you for your detailed post and for using Groonga!

From your post, it seems that you're aiming to achieve a consistent way to search across columns—specifically using ` _key` for exact matches and `title` for full-text searches—without needing to escape user input.  To achieve your goal of consistent searches using exact matches on `_key` and full-text search on `title`, here's a suggestion that might work for you.
- Point: If you want to leverage index-based exact matching, you can do so by not specifying a `default_tokenizer` when creating the index.
  - https://groonga.org/docs/reference/commands/table_create.html#default-tokenizer

```
table_create CouponNames TABLE_PAT_KEY ShortText
table_create Coupons TABLE_HASH_KEY CouponNames
column_create Coupons title COLUMN_SCALAR ShortText
load --table Coupons
[
  {"_key": "fr1", "title": "Orange 10% off"},
  {"_key": "fr10", "title": "Apple 20% off"}
]
column_create CouponNames coupons_key COLUMN_INDEX Coupons _key
table_create Terms TABLE_PAT_KEY ShortText --default_tokenizer TokenBigram --normalizer NormalizerNFKC150
column_create Terms coupons_title COLUMN_INDEX|WITH_POSITION Coupons title
```

With this setup, you can perform searches across both `_key` and `title` using the following queries.
For exact matches on `_key`:
```
select Coupons --match_columns "CouponNames.coupons_key || title" --query fr1
[
  [
    0,
    0.0,
    0.0
  ],
  [
    [
      [
        1
      ],
      [
        [
          "_id",
          "UInt32"
        ],
        [
          "_key",
          "ShortText"
        ],
        [
          "title",
          "ShortText"
        ],
        [
          "coupons_key",
          "Coupons"
        ]
      ],
      [
        1,
        "fr1",
        "Orange 10% off",
        1
      ]
    ]
  ]
]
```
For full-text search on `title`:
```
select Coupons --match_columns "CouponNames.coupons_key || title" --query Apple
[
  [
    0,
    0.0,
    0.0
  ],
  [
    [
      [
        1
      ],
      [
        [
          "_id",
          "UInt32"
        ],
        [
          "_key",
          "ShortText"
        ],
        [
          "title",
          "ShortText"
        ]
      ],
      [
        2,
        "fr10",
        "Apple 20% off"
      ]
    ]
  ]
]
```

I hope this approach helps resolve the issues you're facing. Let me know if you have any further questions or if there are other requirements you'd like to address! Let me know if you'd like to adjust any further details!

## Additional Note
```
table_create CouponNames TABLE_PAT_KEY ShortText
table_create Coupons TABLE_HASH_KEY CouponNames
column_create CouponNames coupons_key COLUMN_INDEX Coupons _key
```
Let me explain the reason for using this reference type (`CouponNames`) in the setup. Using a reference type like this improves space efficiency, as it avoids storing duplicate values. If we don't use this reference and instead directly index `_key`, we would need to store the same values twice—once in the `CouponNames` table and again in the `Coupons` table.
- https://groonga.org/docs/tutorial/data.html#reference-types

Of course, the method below will still work to achieve the desired behavior, but it may be less efficient in terms of storage.
```
table_create Coupons TABLE_HASH_KEY ShortText
table_create CouponNames TABLE_PAT_KEY ShortText
column_create CouponNames coupons_key COLUMN_INDEX Coupons _key
```

**Reply by yssrku** [reply]
Thanks for the detailed explanations, this indeed works!!
So I guess `WITH_POSITION` is not required for the index for Coupons._key

(Side note: I have to add `--match_escalation no` to avoid `fr` from matching)
